File: solutions/julia/queen-attack/1/queen-attack.jl
import Base: ==

struct InvalidPosition <: Exception
    
end

struct Queen
    x::Int8
    y::Int8    
    Queen(r,c) = r<0 || r>=8 || c<0 || c>=8 ?
        throw(InvalidPosition()) : new(r,c)
end

==(w::Queen, b::Queen) = (w.x == b.x && w.y == b.y)

function canattack(white::Queen, black::Queen)
    Δx = abs(white.x - black.x)
    Δy = abs(white.y - black.y)
    Δx == 0 ||  Δy == 0 ||  Δx  ==  Δy 
end    

File: solutions/cpp/queen-attack/1/queen_attack.cpp
#include "queen_attack.h"
#include <utility>
#include <cmath>
#include <stdexcept>

namespace queen_attack {

   bool chess_board::on_board(std::pair<int,int> queen){
      return (  (queen.first>=0) && (queen.first<8) &&
            (queen.second>=0) && (queen.second<8) );
   }

    chess_board::chess_board(std::pair<int,int> w, std::pair<int,int> b): w_q{w}, b_q{b}{
        if  (w == b)
            throw std::domain_error("Queen positions must be distinct.");  
        else if (!on_board(w) || !on_board(b))
            throw std::domain_error("Position not on board."); 
    }

    bool chess_board::can_attack() const {
        int dx = abs(w_q.first - b_q.first);
        int dy = abs(w_q.second - b_q.second);
        return ((!dx) || (!dy) || (dx == dy));
    }
    

}  // namespace queen_attack
